test/functional/rpc_blockchain.py

```diff
@@ -227,6 +240,47 @@ def _test_stopatheight(self):
         self.start_node(0)
         assert_equal(self.nodes[0].getblockcount(), 207)
 
+    def _test_waitforblockheight(self):
+        node = self.nodes[0]
+
+        # Start a P2P connection since we'll need to create some blocks.
+        node.add_p2p_connection(P2PInterface())
+        network_thread_start()
+        node.p2p.wait_for_verack()
+
+        current_height = node.getblock(node.getbestblockhash())['height']
+
+        # Create a fork somewhere below our current height, invalidate the tip
+        # of that fork, and then ensure that waitforblockheight still
+        # works as expected.
+        #
+        # (Previously this was broken based on setting
+        # `rpc/blockchain.cpp:latestblock` incorrectly.)
+        #
+        b20hash = node.getblockhash(20)
+        b20 = node.getblock(b20hash)
+
+        b21f = create_block(
+            int(b20hash, 16), create_coinbase(21), b20['time'] + 1)
+        b21f.solve()
+        node.p2p.send_message(msg_block(b21f))
+        b22f = create_block(
+            b21f.sha256, create_coinbase(22), b21f.nTime + 1)
+        b22f.solve()
+        node.p2p.send_message(msg_block(b22f))
+        node.p2p.sync_with_ping()
+
+        node.invalidateblock(b22f.hash)
+
+        def assert_waitforheight(height, timeout=2):
+            assert_equal(
+                node.waitforblockheight(height, timeout)['height'],
+                current_height)
+
+        assert_waitforheight(0)
+        assert_waitforheight(current_height - 3)
+        assert_waitforheight(current_height)
```

Comment:
Might also want to add a timeout case like in `_test_stopatheight` above.

Reply:
Played around with doing this, but found that the return from RPC is (unfortunately) no different in the timeout case, meaning we'd have to spin up a separate thread in-test, run a timer, and then assert that the RPC thread hadn't yet stopped execution -- probably overkill.

Reply:
I've added a testcase for `current_height + 1`, though.